# Teiler und Vielfache
## Vorbemerkungen zu den Notebooks
In den Übungsblättern (*Jupyter Notebooks*) werden wir die Inhalte der Lektionen einüben, um sicher im Umgang mit den mathematischen Regeln zu werden.
Als Werkzeug hierfür verwenden wir *Python*, bzw. kleinere Programme, die du selbst hier in Python erstellen wirst.

Wir werden hierzu auf Problemstellungen zurückgreifen, die im Internet frei verfügbar sind, z.B. auf den Plattformen [Exercism](https://exercism.org) und [Poject Euler](https://projecteuler.net/).
Du kannst dich direkt dort umsehen und auch gleich einen eigenenen Account anlegen, um deine eigenen Lösungen dort zu prüfen.
Das ist aber nicht unbedingt notwendig, du kannst auch einfach nur in diesem Notebook arbeiten.

Wenn du noch keine oder geringe Programmiererfahrung hast, mach die keine Sorgen, du kannst trotzdem gleich mit der ersten Problemstellung beginnen.
Ich werde dich Schritt für Schritt bis zur Lösung führen.

### Problem 1: [Euler 1](https://projecteuler.net/problem=1)
**Vielfache von 3 oder 5**\
Wenn wir alle natürlicher Zahlen kleiner als 10 auflisten, die Vielfache von 3 oder 5 sind, erhalten wir $3,5,6 \ \mathrm{und} \ 9$.
Die Summe dieser Vielfachen ist $23$.\
**Finde die Summe aller Vielfachen von 3 oder 5 unter 1000**.

Die Lösung eines Problems ist typischerweise nicht sofort ersichtlich; wenn sie das wäre, dann hätten wir gar kein Problem.

Das Problem besteht hier darin, dass wir die Vielfachen von 3 und 5 nicht kennen, und dass es ziemlich viele von ihnen gibt.
Wir suchen also nach einer Formel, wie wir diese Vielfachen einfach berechnen können, um uns manuelle Rechenarbeit zu ersparen.

Eine Zahl $n$ ist ein Vielfaches einer Zahl $m$, wenn $m$ ein Teiler von $n$ ist, sich $n$ also ohne Rest durch $m$ teilen läßt.
Das erinnert an die *Division mit Rest* aus der Lektion: $n \div m = q + r$, wobei $q$ der Ganzzahlquotient und $r$ der Rest der Division ist.
Wenn der Rest der Division $n\div m$ also gleich 0 ist, dann ist $m$ ein Teiler von $n$, und $n$ ist ein Vielfaches von $m$.

Die Teiler $m$ kennen wir: das sind die Zahlen 3 und 5 aus der Problemstellung.
Den Rest der Division $n\div m$ schreiben wir als: $n \ \mathrm{mod} \ m$.
Damit ist eine Zahl $n$ ein Vielfaches von 3 oder 5 wenn gilt: $n \ \mathrm{mod} \ 3=0 \ \mathrm{oder} \ n \ \mathrm{mod} \ 5=0$.

Unser Lösungsansatz könnte also so aussehen:
Wir durchlaufen alle natürlichen Zahlen $n<1000$ und testen, ob 3 oder 5 ein Teiler von $n$ ist.
Wenn ja, dann ist $n$ ein Vielfaches von 3 oder 5, und wir addieren $n$ zu einer laufenden Summe.

Wir wollen unsere Lösungen immer als Funktionen schreiben, damit wir sie später einfach aufrufen können.\
**Aufgabe**: vervollständige die Funktionsdefinition `your_solution` in der folgenden Zelle.\
**Hinweis**: der *modulo* Operator in Python ist `%`. Der Test auf Gleichheit erfolgt mit `==`. 

Zum Öffen einer Zelle wähle sie mit den Cursor-Tasten ↑ oder ↓ aus (oder klicke in die Zelle) und drücke dann die Eingabetaste `<Enter>` (↲). Um den Python Code einer Zelle auszuführen, halte die Umschalttaste gedrückt und drücke die Eingabeteaste: `<shift><Enter>` (⇑↲).\
Achte darauf, jede Code Zelle mindestens einmal auszuführen bevor du zur nächsten Zelle gehst, auch wenn du gar keine Änderung vorgenommen hast.

In [1]:
def your_solution(limit):
    n = 1
    res = 0
    # definiere hier deine while Schleife:
    
    return res

# ändere nicht die folgende Zeile
assert your_solution(10) == 23, "das Ergebnis ist nicht korrekt"

AssertionError: das Ergebnis ist nicht korrekt

Die letzte Zeile der obigen Zelle testet, ob dein Ergebnis korrekt ist.
Wir verwenden zum Test die Information aus der Problemstellung: die Summe der Vielfachen von 3 oder 5 kleiner als 10 ist gleich 23.

Wenn du bei der Ausführung der Zelle einen `AssertionError` erhältst, dann ist dein Ergebnis nicht richtig.
Wenn das passiert, finde deinen Fehler und behebe ihn.
Falls du nicht weiterkommst, sieh dir meine Lösung in der nächsten Zelle an.

Wenn dein Programm nicht mehr reagiert, dann hast du möglicherweise eine Endlos-Schleife erzeugt (vielleicht hast du vergessen, die Schleifenvariable `n` hochzuzählen?).
In diesem Fall stoppe die Programmausführung mit dem Befehl `Interrupt Kernel` aus dem Menüpunkt `Kernel` von Jupyter.
Du kannst hierzu auch auf das kleine graue Quadrat in der Symbolleiste des Notebooks klicken.

In [2]:
def my_solution(limit):
    n = 1
    res = 0
    while n < limit:
        if n % 3 == 0 or n % 5 == 0:
            res += n
            # debugging info:
            print(f"n: {n}, result: {res}")
        n += 1
    return res

assert my_solution(10) == 23

n: 3, result: 3
n: 5, result: 8
n: 6, result: 14
n: 9, result: 23


Ich habe hier innerhalb der Schleife eine Zeile `debug info` eingefügt, die den aktuellen Wert der Variablen `n` und `res` für jeden Durchlauf ausgibt.
Die Ausgaben der `debug info` können sehr hilfreich sein, um Fehler in deiner Programmlogik zu entdecken.

Das Programm arbeitet offensichtlich richtig, zumindest für das `limit` 10.
Das bietet aber keine Gewähr, dass es auch für das `limit` von 1000 korrekt arbeitet; es könnte einen Fehler enthalten, der sich erst für größere Argumente auswirkt.

Da wir das korrekte Ergebnis für $limit=1000$ aber nicht kennen, haben wir zunächst keine Möglichkeit, weitere Tests durchzuführen.\
Wir wenden daher folgende Strategie an:
wir erstellen eine zweite Version der Lösung, die einen anderen Algorithmus verwendet, und vergleichen dann die Ergebnisse beider Funktionen.

Für die neue Version verwenden wir eine sogenannte *list comprehension*: die Werte der Vielfachen werden einer Liste hinzugefügt, indem wir eine `for` Schleife innerhalb von eckigen Klammern (begrenzt durch `[` und `]`) definieren.
Die Funktion `sum()` berechnet dann die Summe der Listen-Elemente:

In [3]:
def solve_with_comprehension(limit):
    return sum([n for n in range(1, limit) if n%3 == 0 or n%5 == 0])

# Neudefinition von my_solution ohne debug info:
def solve(limit):
    n = 1
    res = 0
    while n < limit:
        if n % 3 == 0 or n % 5 == 0:
            res += n
        n += 1
    return res

assert solve_with_comprehension(10) == 23

count = 0
for limit in range(10, 1001, 11):
    count += 1
    assert solve(limit) == solve_with_comprehension(limit)
count

91

Unsere neue Funktion liefert das gleiche Ergebnis wie unsere erste Lösung.
Um die Testsabdeckung zu verbessern, habe ich mehrere Tests in einer Schleife ausgeführt, wobei jeweils ein anderer Wert für `limit` gesetzt wird.
Das dritte Argument der `range()` Funktion gibt dabei die Schrittweite an; wir haben unsere Funktionen also mit den Argumenten $10, 21, 32, \dots, 1000$ aufgerufen, insgesamt 91 mal.

Damit können wir ziemlich sicher sein, dass unser Ergebnis für `solve(1000)` korrekt ist, und das Problem ist gelöst.
Wir sind aber noch nicht fertig: 
die Lösung ist zwar korrekt, aber wir können sie noch verbessern, indem wir nach einer effizienteren Lösung suchen.

Der Ausgangspunkt für eine efizientere Lösung ist die folgedne Idee:
anstatt zu prüfen, ob $n$ durch 3 oder 5 teilbar ist, könnten wir zunächst alle Vielfachen von 3 und dann die Vielfachen von 5 suchen, und schließlich die Ergebnisse addieren.
Dann müssten wir aber die gemeinsamen Vielfachen von 3 und 5 wieder subtrahieren (das sind die Vielfachen von $3\cdot5=15$), da wir diese doppelt addiert hätten.

Dann würden die beiden Teilsummen für die Vielfachen von 3 und 5 so aussehen
$$
\begin{align}
3 + 6 + 9 + \cdots + 999 &= 3*(1 + 2 + 3 +\cdots + 333) \\
5 + 10 + 15 + \cdots + 995 &= 5*(1 + 2 + 3 + \cdots + 199)
\end{align}
$$
und wir könnten die [Gaußsche Summenformel](https://de.wikipedia.org/wiki/Gau%C3%9Fsche_Summenformel) anwenden:
$$
1+2+3+\cdots+n = \sum_{k=1}^n k = \frac{n(n+1)}{2}
$$
Die Summenformel kann für verschiedene Teiler wie folgt in Python implementiert werden:

In [4]:
def gauss(limit, teiler):
    m = (limit-1) // teiler
    summe = (m * (m + 1)) // 2
    return teiler * summe

assert gauss(10, 3) == 3 + 6 + 9
assert gauss(10, 5) == 5

Bleibt nur noch der letzte Schritt: addieren der beiden Teilsummen für 3 und 5, und davon die Summe für 15 subtrahieren.\
**Aufgabe**: vervollständige die Funktionsdefinition `your_optimized`.

In [5]:
def your_optimized(limit):
    return limit

assert your_optimized(1000) == solve(1000), "Die Ergebnisse stimmen nicht überein"

AssertionError: Die Ergebnisse stimmen nicht überein

In [6]:
def my_optimized(limit):
    return gauss(limit, 3) + gauss(limit, 5) - gauss(limit, 15)

assert my_optimized(1000) == solve(1000)

Die optimierte Funktion arbeitet korrekt, aber sie ist nicht besonders elegant:
wir rufen wiederholt `gauss` mit dem immer gleichen *limit* auf.

Um das zu vermeiden, können wir `gauss` als innere Funktion der Lösung definieren und dabei auf den Parameter `limit` verzichten, da dieser Parameter in der äußeren Funktion eingeführt wird.\
**Aufgabe**: definiere die Funktion `your_optimized` neu, indem du `gauss` als innere Funktion definierst.

In [7]:
def your_optimized(limit):
    def gauss(teiler):
        # berechne hier die Summe für den gegebenen Teiler:
        summe = teiler
        
        return summe
    # berechne und gib das Endergebnis zurück:
    return gauss(3)

assert your_optimized(1000) == solve(1000), "Die Ergebnisse stimmen nicht überein"

AssertionError: Die Ergebnisse stimmen nicht überein

In [8]:
def my_optimized(limit):
    def gauss(teiler):
        m = (limit-1) // teiler
        summe = (m * (m + 1)) // 2
        return teiler * summe
    return gauss(3) + gauss(5) - gauss(15)

assert my_optimized(1000) == solve(1000)

Vielleicht fragst du dich, ob sich der Aufwand gelohnt hat und die optimierte Lösung tatsächlich effizienter ist.
Das können wir mit dem *magic command* `%timeit` überprüfen, das die Laufzeit einer Anweisung misst, und uns so einen Einfruck über die Effizienz eines Algorithmus verschafft:

In [9]:
%timeit solve(1000)

44.5 μs ± 324 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [26]:
%timeit my_optimized(1000)

375 ns ± 3.51 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


Die Laufzeit für einen Aufruf von `solve(1000)` beträgt im Durchschnitt 44 Mikrosekunden ($\mu s$).
Eine Mikrosekunde ist eine Millionstel Sekunde ($10^{-6}s$).\
Die Laufzeit für einen Aufruf der optimierten Lösung beträgt im Durchschnitt 375 Nanosekunden (ns).
Eine Nanosekunde ist eine Milliardstel Sekunde ($10^{-9}s$).\
Die optimierte Lösung ist also mehr als 100 mal schneller als unsere erste Lösung.

**Beachte:** Wenn du die Zellen mit Zeitmessung ausführst, bekommst du wahrscheinlich abweichende Ergebnisse, abhängig von der Geschwindigkeit deines Computers.

### Problem 2: [Euler 3](https://projecteuler.net/problem=3)
**Größter Primfaktor**\
Die Primfaktoren von $13195$ sind $5,7,13,29$.\
**Was ist der größte Primfaktor der Zahl $600851475143$?**

Ein Primfaktor einer Zahl $n$ ist eine Zahl, die Teiler von $n$ ist und dabei auch eine Primzahl ist.
Eine Primzahl ist eine Zahl, die nur zwei Teiler hat, 1 und die Zahl selbst. 
Wir können also davon ausgehen, dass die gegebene Zahl $n$ keine Primzahl ist, denn dann wäre ihr größter Primfaktor die Zahl selbst, und wir hätten kein Problem zu lösen.

Allein das Wort *Teiler* erinnert uns an **Problem 1**, und damit an die *Division mit Rest*.
Wir könnten also versuchen, auch dieses Problem mit Hilfe des Modulo Operators `%` zu lösen.
Ein erster Lösungsansatz könnte daher so aussehen:

Wir durchlaufen alle natürlichen Zahlen größer als 1 und testen, ob die aktuelle Zahl $p$ die Eingabezahl $n$ ohne Rest teilt.
Falls $p$ ein Faktor von $n$ ist, dann durchlaufen wir erneut alle natürlichen Zahlen größer 2 und testen dieses mal $p$ auf Teiler.
Falls $p$ einen Teiler hat, dann kann $p$ keine Primzahl sein;
anderenfalls fügen wir $p$ zu einer Liste von Primfaktoren hinzu und geben am Ende diese Liste aus.

**Aufgabe**: implementiere diesen Algorithmus in der Funktion `your_solution` in der folgenden Zelle.\
**Hinweis**: du brauchst eine verschachtelte Schleife, also eine Schleife in einer Schleife, um beide Durchläufe für $n$ und $p$ zu realisieren.

In [32]:
def your_solution(n):
    p = 2
    factors = []
    # implementiere hier eine verschachtelte Schleife
    
    return factors

assert your_solution(13195) == [5, 7, 13, 29], "das Ergebnis ist nicht korrekt"

AssertionError: das Ergebnis ist nicht korrekt

In [33]:
def my_solution(n):
    p = 2
    factors = []
    while p < n:
        if n % p == 0:    # p ist ein Faktor von n
            prime = True
            k = 2
            while k < p:
                if p % k == 0:    # p ist keine Primzahl
                    prime = False
                    break
                k += 1
            if prime: factors.append(p)
        p += 1
    return factors

assert my_solution(13195) == [5, 7, 13, 29]

Beachte die `break` Anweisung in der inneren Schleife: wenn `p` einen Teiler hat, dann ist `p` keine Primzahl, und wir können die Schleife abbrechen, d.h. wir sparen uns den Test auf weitere Teiler.

Wenn wir nun diese Lösung mit $600851475143$ aufrufen stellen wir fest, dass die Funktion kein Ergebnis zurückliefert, zumindest nicht in absehbarer Zeit.
Der Algorithmus ist schlichtweg nicht effizient genug, um die Primfaktoren einer großen Zahl zu berechnen.

In [60]:
%timeit my_solution(13195)

460 μs ± 7.81 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


Hier wurde die Funktion in 7 Durchläufen jeweils 1000 mal aufgerufen, mit einer durchschnittlichen Laufzeit von 453 $\mu s$ pro Aufruf.
Das erscheint erst einmal nicht besonders viel, aber verschachtelte Schleifen haben es in sich:
die innere Schleife wird für jedes Element der äußeren Schleife ausgeführt.
Wenn wir also $n$ Elemente in der äußeren und $m$ Elemente in der inneren Schleife durchlaufen, haben wir insgesamt $n\cdot m$ Operationen.
Da wir nicht genau wissen, wie viele Elemente die innere Schleife hat, gehen wir bei der Betrachtung der Laufzeiten von Algorithmen vom ungünstigsten Fall aus.

Im ungünstigsten Fall hat die innere Schleife ebenfalls $n$ Elemente; die gesamte Laufzeit beider Schleifen ist dann $n \cdot n=n^2$.
Wir sprechen deshalb von einer *quadratischen Laufzeit*.\
Bei der Betrachtung von Laufzeiten hat sich die sogenannte *big O notation* durchgesetzt.
Das große O steht dabei für *Ordnung*, im Sinne von *Größenordnung*.
Die Laufzeit unseres Algorithmus liegt damit in der Größenordnung von $n^2$, wobei $n$ die Größe des Eingabewerts bezeichnet.
Das wird dann geschrieben als $O(n^2)$.


Wie können wir jetzt aber die Laufzeit unseres Algorithmus verbessern?\
Bisher haben wir immer nur die Teilbarkeit von $n$ betrachtet, die Teiler selbst aber unberücksichtigt gelassen.
Was wäre aber, wenn wir $n$ nun tatsächlich durch die gefundenen Teiler dividieren würden?\
Dann könnten wir mit dem neuen $n$ weiterrechnen und innerhalb der Schleife nach weiteren Teilern von $n$ suchen, und $n$ durch jeden dieser Teiler dividieren.

Wenn wir dabei jeden gefunden Teiler $p$ komplett ausdividieren, d.h. wenn wir $n$ solange durch $p$ teilen, wie das neue $n$ eben durch $p$ teilbar ist, dann ist jeder neue Teiler automatisch eine Primzahl.
Das folgt aus der Tatsache, dass alle Faktoren, die keine Primzahl sind, also selbst als Produkt ihrer Primfaktoren dargestellt werden können, bereits ausdividiert sind.
Das bedeutet auch, wir sparen uns den Test, ob $p$ eine Primzahl ist.\
Wenn $n$ gleich 1 ist, dann sind alle Faktoren ausdividiert, und wir geben $p$ als letzten und damit größten Primfaktor als Ergebnis zurück.

**Aufgabe**: implementiere diesen Algorithmus in der Funktion `my_optimized` in der folgenden Zelle.\
**Hinweis**: du brauchst auch hier eine verschachtelte Schleife, und zwar für das vollständige Ausdividieren jedes gefundenen Teilers in der inneren Schleife.

In [34]:
def your_optimized(n):
    p = 2
    # implementiere hier deine Schleife:
    
    return p

assert your_optimized(13195) == 29, "das Ergebnis ist nicht korrekt"

AssertionError: das Ergebnis ist nicht korrekt

In [35]:
def my_optimized(n):
    p = 2
    while p <= n:
        while n % p == 0:
            n //= p
        if n == 1:
            return p
        else:
            p += 1

assert my_optimized(13195) == 29

In [36]:
%timeit my_optimized(13195)

951 ns ± 3.91 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


Die durchschnittliche Laufzeit ist jetzt 951 Nanosekunden (*ns*).
Damit ist der optimierte Algorithmus ungefähr 500 mal schneller als unsere bisherige Lösung.

Wir können aber noch weiter optimieren:
es ist nämlich gar nicht notwendig, alle natürlichen Zahlen bis $n$ zu durchlaufen.
Es reicht, die Zahlen bis zur Obergrenze von $\sqrt n$ zu durchlaufen, da jede Zahl $n$ höchstens *einen* Primfaktor größer als $\sqrt n$ haben kann.\
Diese Obergrenze können wir anstatt mit `p <= sqrt(n)` auch mit `p*p <= n` angeben, da $p=\sqrt n$ gleichbedeutend ist mit $p^2=n$.
Das erspart einige Rechenoperationen zur Berechnung der Wurzel.

Wenn wir das tun, dann müssen wir zwei Fälle unterscheiden:
entweder ist nach der letzten Divison $n$ gleich 1, dann geben wir wie bisher $p$ als größten Primfaktor zurück.
Oder $n$ ist ungleich 1, dann geben wir $n$ zurück, denn dann ist das aktuelle $n$ der *eine* Primfaktor größer als $\sqrt n$.

In [37]:
def my_optimized(n):
    p = 2
    while p*p <= n:
        while n%p == 0:
            n //= p
        if n == 1:
            return p
        p += 1
    return n

assert my_optimized(13195) == 29

In [39]:
%timeit my_optimized(13195)

560 ns ± 10.4 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


Die durchschnittliche Laufzeit ist jetzt 560 Nanosekunden, also knapp doppelt so schnell wie zuvor.
Wir können die optimierte Lösung jetzt also getrost mit dem Argument $600851475143$ aufrufen und verwenden für die Verifizierung des Ergebnisses deine und meine Version (du solltest deine Version `your_optimized` also bereits fertiggestellt haben):

In [66]:
assert my_optimized(600851475143) == your_optimized(600851475143), "die Ergebnisse stimmen nicht überein"

Abschließend kannst du noch die Laufzeiten unserer Versionen vergleichen:

In [67]:
%timeit my_optimized(600851475143)

85.5 μs ± 1.3 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [68]:
%timeit your_optimized(600851475143)

288 μs ± 1.79 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


### Problem 3: [Euler 5](https://projecteuler.net/problem=5)
**Kleinstes Vielfaches**\
$2520$ ist die kleinste Zahhl, die sich ohne Rest durch alle Zahlen von 1 bis 10 teilen läßt.\
**Was ist die kleinste positive Zahl, die sich ohne Rest durch alle Zahlen von 1 bis 20 teilen läßt?**



Aus der Problemstellung können wir ohne Weiteres schließen, dass wir es wieder mit Teilern und Vielfachen zu tun haben.
Außerdem erinnert der Titel des Problems *kleinstes Vielfaches* uns an den Begriff *kleinstes gemeinsames Vielfaches* (kgV).

Nach kurzem Nachdenken kommen wir zu der Erkenntnis, dass in diesem Problem tatsächlich nach dem *kgV* der Zahlen 1 bis 20 gesucht wird, denn das *kgV* ist wie folgt definiert: *die kleinste positive Zahl, die ein Vielfaches der genannten Zahlen ist*, also die kleinste positive Zahl, die sich ohne Rest durch diese Zahlen läßt.

Das *kgV* für mehrere Zahlen läßt sich über die *Primfaktorzerlegung* dieser Zahlen berechnen:
wir multiplizieren alle Primfaktoren, die in mindestens einer der Zahlen vorkommen, in ihrer jeweils höchsten vorkommenden Potenz.

Die höchsten Potenzen der Primfaktoren der Zahlen 1 bis 10 sind:
* $2^3=8$
* $3^2=9$
* $5^1=5$
* $7^1=7$

Damit können wir das *kgV* der Zahlen 1 bis 10 so berechnen:

In [85]:
from math import prod

kgV_10 = prod([8, 9, 5, 7])
assert kgV_10 == 2520

Hier haben wir die Funktion `prod` aus dem Python Modul `math` importiert, um direkt auf sie zugreifen zu können.
Diese Funktion berechnet das Produkt der Zahlen, die als sogenanntes *iterable* angegeben werden, also z.B. als Liste.
**Aufgabe**: berechne nach diesem Muster das *kgV* der Zahlen 1 bis 20.

In [55]:
kgV_20 = 1

assert kgV_20 == 232792560, "das Ergebnis ist nicht korrekt"

AssertionError: das Ergebnis ist nicht korrekt

Die höchsten Potenz der Primfaktoren größer als 3 ist jeweils 1, da bereits $5^2=25>20$.
Die höchste Potenz von 2 ist jetzt aber 4, da $2^4=16\leq20$.\
Wir können das *kgV* der Zahlen 1 bis 20 damit so berechnen: 

In [86]:
kgV_20 = prod([16, 9, 5, 7, 11, 13, 17, 19])

Damit ist das Problem gelöst, aber als angehende Entwickler wollen wir uns damit nicht zufrieden geben.
Wie in den vorangehenden Problemen wollen wir eine Funktion erstellen, die das *kgV* für die ersten $n$ natürlichen Zahlen berechnet.

Dazu müssen wir zunächst jede Zahl in ihre Primfaktoren zerlegen.
Wir verwenden dazu den Algorithmus der Funktion `my_optimized` aus **Problem 2**, nur dass wir jetzt eine zusätzliche Liste `factors` einführen, in der wir alle gefundenen Primfaktoren speichern, und am Ende diese Liste ausgeben.\
**Aufgabe**: Implementiere eine Funktion `factorize`, die eine Zahl nach diesem Muster in ihre Primfaktoren zerlegt.

In [37]:
def factorize(n):
    factors = []
    
    return factors

assert factorize(18) == [2,3,3], "das Ergebnis ist nicht korrekt"

AssertionError: das Ergebnis ist nicht korrekt

In [38]:
def my_factorize(n):
    factors = []
    p = 2
    while p*p <= n:
        while n%p == 0:
            factors.append(p)
            n //= p
        if n == 1:
            return factors
        p += 1
    factors.append(n)
    return factors

assert my_factorize(18) == [2,3,3]

Als nächstes müssen wir für jeden gefundenen Primfaktor einer Zahl die jeweilige Potenz bestimmen.
Dazu zählen wir ab, wie oft ein Faktor in einer Liste von Faktoren vorkommt und speichern das Ergebnis in einem *dictionary*, einem Objekt mit dem Datentyp `dict`.

Ein `dict` ist ein Objekt, dessen Elemente sogenannte *key-value pairs* sind, also Paare von Schlüsseln und zugehörigen Werten.
In diesem Fall bilden die Faktoren die *keys*, und die zugehörigen *values* sind ihre jeweils höchsten Potenzen.\
**Aufgabe**: Implementiere eine Funktion `powers`, die die Faktoren einer Liste abzählt und das Ergebnis als *dictionary* zurückgibt.\
**Hinweis**: Verwende für den Zugriff auf einen *key* `k` die Funtion `dict.get(k, 0)`. Falls `k` nicht im *dictionary* enthalten ist, wird der *default value* 0 zurückgegeben.

In [44]:
def powers(factors):
    hp =  {}    # ein leeres dictionary
    # implementiere hier eine for-Schleife, die alle Elemente von factors durchläuft:
    
    return hp

assert powers([2,3,3]) == {2: 1, 3: 2}, "das Ergebnis ist nicht korrekt"

AssertionError: das Ergebnis ist nicht korrekt

In [63]:
def my_powers(factors):
    hp = {}
    for factor in factors:
        # Hochzählen des Potenz, indem wir den aktuellen Wert um eins erhöhen
        hp[factor] = hp.get(factor, 0) + 1
    return hp

assert my_powers([2,3,3]) == {2: 1, 3: 2}

Um jetzt die jeweils höchsten Potenzen der Primfaktoren zweier Zahlen zu finden, vergleichen wir wir ihre Potenzen und führen sie in einem neuen *dictionary* zusammen.\
**Aufgabe**: Implementiere eine Funktion `highest_powers`, die zwei *dictionaries* als Argumente akzeptiert und ein neues *dictionary* mit den jeweils höchsten Potenzen zurückgibt.\
**Hinweis**: Verwende für das Zusammenführen zweier *dictionaries* den Operator `|`.

In [69]:
def highest_powers(d1, d2):
    merged = d1
    
    return merged

# ändere nicht den nachfolgenden Code
d1 = my_powers([2,3,3])
d2 = my_powers([2,2,5])
assert highest_powers(d1, d2) == {2: 2, 3: 2, 5: 1}, "das Ergebnis ist nicht korrekt"

AssertionError: das Ergebnis ist nicht korrekt

In [71]:
def my_highest_powers(d1, d2):
    merged = d1 | d2
    for k in merged.keys():
        merged[k] = max(d1.get(k, 0), d2.get(k, 0))
    return merged

assert my_highest_powers(d1, d2) == {2: 2, 3: 2, 5: 1}

Schließlich können wir die höchsten Potenzen der Primfaktoren aller Zahlen $\leq n$ berechnen, indem wir paarweise die Funktion `highest_powers` für ihre Primfaktorzerlegungen aufrufen.\
**Aufgabe**: Implementiere eine Funktion `hp_all`, die eine natürliche Zahl $n>1$ als Argument nimmt und ein *dictionary* mit den höchsten Potenzen aller Primfaktoren der Zahlen $\leq n$ zurückliefert.\
**Hinweis**: Verwende dazu deine Funktionen `factorize`, `powers` und `highest_powers`.

In [81]:
def hp_all(n):
    res = {2: 1}    # Initialisierung des Ergebnisses mit der Primfaktorzerlegung von 2
    
    return res

assert hp_all(10) == {2: 3, 3: 2, 5: 1, 7: 1}, "das Ergebnis ist nicht korrekt"

AssertionError: das Ergebnis ist nicht korrekt

In [82]:
def my_hp_all(n):
    res = {2: 1}
    for i in range(3, n+1):
        res = my_highest_powers(res, my_powers(my_factorize(i)))
    return res

assert my_hp_all(2) == {2: 1}
assert my_hp_all(10) == {2: 3, 3: 2, 5: 1, 7: 1}
assert my_hp_all(20) == {2: 4, 3: 2, 5: 1, 7: 1, 11: 1, 13: 1, 17: 1, 19: 1}
assert my_hp_all(30) == {2: 4, 3: 3, 5: 2, 7: 1, 11: 1, 13: 1, 17: 1, 19: 1, 23: 1, 29: 1}

Im letzten Schritt der Lösung berechnen wir das Produkt der höchsten Potenzen aller Primfaktoren der Zahlen $\leq n$.\
**Aufgabe**: Implementiere deine Lösung in der Funktion `your_solution`.

In [91]:
def your_solution(n):
    res = 1
    
    return res

assert your_solution(10) == kgV_10, "das Ergebnis ist nicht korrekt"
assert your_solution(20) == kgV_20, "das Ergebnis ist nicht korrekt"

AssertionError: das Ergebnis ist nicht korrekt

In [92]:
def my_solution(n):
    res = 1
    hpf = my_hp_all(n)
    for (k, v) in hpf.items():
        res *= k ** v
    return res

assert my_solution(10) == kgV_10
assert my_solution(20) == kgV_20